drop-transfer/src/ws/server/v3.rs

```diff
@@ -0,0 +1,535 @@
+use std::{
+    collections::HashMap,
+    fs,
+    net::IpAddr,
+    ops::ControlFlow,
+    path::PathBuf,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use anyhow::Context;
+use drop_config::DropConfig;
+use futures::{SinkExt, StreamExt};
+use slog::{debug, error, info, warn};
+use tokio::{
+    sync::{
+        mpsc::{self, Sender, UnboundedSender},
+        oneshot,
+    },
+    task::JoinHandle,
+};
+use warp::ws::{Message, WebSocket};
+
+use super::{handler, ServerReq};
+use crate::{protocol::v3, service::State, utils::Hidden, ws::events::FileEventTx, FileId};
+
+pub struct HandlerInit<'a> {
+    peer: IpAddr,
+    state: &'a Arc<State>,
+    logger: &'a slog::Logger,
+}
+
+pub struct HandlerLoop<'a> {
+    state: &'a Arc<State>,
+    logger: &'a slog::Logger,
+    msg_tx: Sender<Message>,
+    xfer: crate::Transfer,
+    last_recv: Instant,
+    jobs: HashMap<FileId, FileTask>,
+}
+
+struct Downloader {
+    logger: slog::Logger,
+    file_id: crate::FileId,
+    msg_tx: Sender<Message>,
+    tmp_loc: Option<Hidden<PathBuf>>,
+    csum_rx: Option<oneshot::Receiver<v3::ReportChsum>>,
+    offset: u64,
+}
+
+struct FileTask {
+    job: JoinHandle<()>,
+    chunks_tx: UnboundedSender<Vec<u8>>,
+    events: Arc<FileEventTx>,
+    csum_tx: Option<oneshot::Sender<v3::ReportChsum>>,
+}
+
+impl<'a> HandlerInit<'a> {
+    pub(crate) fn new(peer: IpAddr, state: &'a Arc<State>, logger: &'a slog::Logger) -> Self {
+        Self {
+            peer,
+            state,
+            logger,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<'a> handler::HandlerInit for HandlerInit<'a> {
+    type Request = (v3::TransferRequest, IpAddr, DropConfig);
+    type Loop = HandlerLoop<'a>;
+    type Pinger = tokio::time::Interval;
+
+    async fn recv_req(&mut self, ws: &mut WebSocket) -> anyhow::Result<Self::Request> {
+        let msg = ws
+            .next()
+            .await
+            .context("Did not received transfer request")?
+            .context("Failed to receive transfer request")?;
+
+        let msg = msg.to_str().ok().context("Expected JOSN message")?;
+
+        let req = serde_json::from_str(msg).context("Failed to deserialize transfer request")?;
+
+        Ok((req, self.peer, self.state.config))
+    }
+
+    async fn on_error(&mut self, ws: &mut WebSocket, err: anyhow::Error) -> anyhow::Result<()> {
+        let msg = v3::ServerMsg::Error(v3::Error {
+            file: None,
+            msg: err.to_string(),
+        });
+
+        ws.send(Message::from(&msg))
+            .await
+            .context("Failed to send error message")?;
+        Ok(())
+    }
+
+    fn upgrade(self, msg_tx: Sender<Message>, xfer: crate::Transfer) -> Self::Loop {
+        let Self {
```

Comment:
Destructurize `Self` aka `HandlerInit` into fields. Ignore `peer` with `_` since we're not using it and clippy would warn about unused variable without that